### PR TITLE
HPCC-24635 Add NetworkPolicy to helm charts for ingress control

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@jakesmith Please review

File: helm/hpcc/templates/network.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: default-deny-all
+spec:
+  podSelector: {}
+  policyTypes:
+  - Ingress
+  - Egress
+  egress:
```
Comment (richardkchapman) [inline helm/hpcc/templates/network.yaml:11]: You need to allow egress to kube-system for dns lookup to work, However this method is not very reliable, as it relies on the label having been set on the kube-system namespace which it is not by default.

Alternatives are:

1. Don't add egress controls
2. Allow egress to any machine on port 53 (dns)

File: helm/hpcc/templates/network.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+---
+kind: NetworkPolicy
+apiVersion: networking.k8s.io/v1 
+metadata:
+  name: allow-apiserver
+spec:
+  policyTypes:
+  - Egress
+  podSelector: {}
+  egress:
```
Comment (richardkchapman) [inline helm/hpcc/templates/network.yaml:26]: Need to allow egress to the server used by kubectl, for components that use it. However the cidr to specify will vary from system to system.

Alternatives:
1. Don't block egress by default
2. Allow helm user to specify kubectl server to unblock, if they want to block egress

**Comment by jakesmith** [comment]
```
[ERROR] templates/: template: hpcc/templates/thor.yaml:40:19: executing "hpcc/templates/thor.yaml" at <include "hpcc.daliArg" $>: error calling include: template: hpcc/templates/_helpers.tpl:364:19: executing "hpcc.daliArg" at <index .Values.dali 0>: error calling index: reflect: slice index out of range
```

@richardkchapman - is that new?

File: common/workunit/workunit.cpp [hunk 3]
```diff
@@ -14100,13 +14100,23 @@ void waitK8sJob(const char *componentName, const char *job)
...
 }
 
-void launchK8sJob(const char *componentName, const char *wuid, const char *job, const std::list<std::pair<std::string, std::string>> &extraParams)
+bool applyK8sYaml(const char *componentName, const char *wuid, const char *job, const char *suffix, const std::list<std::pair<std::string, std::string>> &extraParams, bool optional)
 {
-    VStringBuffer jobname("%s-%s", componentName, job);
+    StringBuffer jobname(job);
     jobname.toLowerCase();
-    VStringBuffer jobSpecFilename("/etc/config/%s-jobspec.yaml", componentName);
+    VStringBuffer jobSpecFilename("/etc/config/%s-%s.yaml", componentName, suffix);
```
Comment (jakesmith) [inline common/workunit/workunit.cpp:14107]: pedantic: perhaps should remove 'job' from here and jobYaml below, to become k8sSpecFilename / k8sYaml

**Comment by richardkchapman** [comment]
I think that's from one of the tests that's supposed to fail

File: helm/hpcc/templates/network.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  podSelector: {}
+  policyTypes:
+  - Ingress
+  - Egress
+  egress:
+  - to:
+#  This only works if we know that the kube-system namespace has been labelled, which we cannot control?
+    - namespaceSelector:
+        matchLabels:
+          name: kube-system
```
Comment (jakesmith) [inline helm/hpcc/templates/network.yaml:16]: so if the name space of the kube-system is not "kube-system" this NP will not be in force?
In other words, if name space doesn't match - this deny-all NP is not used?

**Comment by jakesmith** [comment]
@richardkchapman - all looks sensible, a couple of questions.

File: helm/hpcc/templates/network.yaml (hunk 4 above)
Comment (richardkchapman) [reply helm/hpcc/templates/network.yaml:16]: This isn't a deny - it's an allow... meaning if it is not applied no-one can call kubectl

And it's not the name of the namespace that matters but its label, which is not present by default.

**Comment by richardkchapman** [comment]
@ghalliday

**Comment by ghalliday** [comment]
@richardkchapman needs squashing and rebasing, otherwise looks ok.

**Pull request merged**